Public/Src/Tools/Tool.MsBuildGraphBuilder/MsBuildGraphBuilder.cs

```diff
@@ -360,31 +357,20 @@ private static ProjectGraphWithPredictionsResult<string> CreateFailureFromInvali
                 ProjectInstance projectInstance = msBuildNode.ProjectInstance;
                 Project project = projectInstanceToProjectCache[projectInstance];
 
-                StaticPredictions predictions;
+                var inputFilePredictions = new List<string>();
+                var outputFolderPredictions = new List<string>();
+
+                var predictionCollector = new MsBuildPredictionCollector(inputFilePredictions, outputFolderPredictions, predictionFailures);
                 try
                 {
                     // Again, be defensive when using arbitrary predictors
-                    predictions = predictionExecutor.PredictInputsAndOutputs(project);
+                    predictionExecutor.PredictInputsAndOutputs(project, predictionCollector);

```

Comment:
>PredictInputsAndOutputs [](start = 39, length = 23)

I haven't seen what this is doing with the collector, but I'd assume it does equivalent logic that the code you removed here, right?

Reply:
Correct. The collector stuff just alleviates the need to map MSBuild Prediction data objects back to objects the engine cares about, saving allocations.

In essence, the `TryGetInputPredictions` and `TryGetOutputPredictions` was moved to the collector to be called directly by the prediction library.

---
In reply to: [294436161](https://github.com/microsoft/BuildXL/pull/480#discussion_r294436161) [](ancestors = 294436161)